src/protocolsupport/protocol/packet/middleimpl/serverbound/play/v_pe/CustomPayload.java

```diff
@@ -2,11 +2,11 @@
 
 import io.netty.buffer.ByteBuf;
 import protocolsupport.protocol.ConnectionImpl;
-import protocolsupport.protocol.packet.middle.serverbound.play.MiddleCustomPayload;
 import protocolsupport.protocol.serializer.MiscSerializer;
 import protocolsupport.protocol.serializer.StringSerializer;
 
-public class CustomPayload extends MiddleCustomPayload {
+//TODO: this really shouldnt need to extend the legacy key name fix version...
```

Comment:
Indeed. Is it possible to move the common functionality to a more "generic" base class?

Reply:
probably. ive been hesitant about making it any more 'permanent' cause i still dont understand why i need it... but, it does work. maybe ill make the correct class for it and slap a big TODO/NOTE on it 🤷‍♂️

Reply:
Please explain this change. Why move away from MiddleCustomPayload?

Reply:
i still need the logic thats in protocolsupport.protocol.packet.middleimpl.serverbound.play.v_8_9r1_9r2_10_11_12r1_12r2.CustomPayload.

I was using it directly before, and on your last commits you moved it to a normal MiddleCustomPayload. For whatever reason i still can't explain, I need the 'legacy key' support in the v_8_9r1_9r2_10_11_12r1_12r2 CustomPayload otherwise the bungee message keys come in wrong. This does need to be cleaned up, but again im a bit wary about integrating it more permanently when im still not totally sure why i need it :(

Reply:
Wasn't there something with old keys inside the bungee side you said worked in a weird way? Is that related to this?

Reply:
i think they're both artifacts of the same issue. i honestly think it has something to do with the hacks in waterfall to get old clients to work? so not only is there key conversion in PSPE, but also PS bungee, and waterfall itself :(